queue/back_pressure_test.go

```diff
@@ -0,0 +1,375 @@
+		capacity := rapid.IntRange(0, 20).Draw(rt, "capacity")
+
+		// The predicate doesn't matter much here as the queue will be
+		// empty.
+		q := NewBackpressureQueue[int](
+			capacity, RandomEarlyDrop[int](0, capacity),
+		)
+
+		require.Zero(
+			rt, len(q.ch), "queue should be empty initially for "+
+				"Dequeue cancellation test",
+		)
+
+		// Attempt to dequeue from the empty queue with an immediately
+		// cancelled context.
+		ctx, cancel := context.WithCancel(context.Background())
+		cancel()
+
+		_, err := q.Dequeue(ctx)
+		require.Error(
+			rt, err, "dequeue should have returned an error "+
+				"for cancelled context",
+		)
+		require.ErrorIs(
+			rt, err, context.Canceled, "Error should be "+
+				"context.Canceled",
+		)
+	})
+}
+
+// TestBackpressureQueueComposedPredicate demonstrates testing with a composed
+// predicate. This is a scenario-based test rather than a full property-based
+// state machine.
+func TestBackpressureQueueComposedPredicate(t *testing.T) {
+	capacity := 10
+	minThresh, maxThresh := 3, 7
+
+	// Use a deterministic random source for this specific test case to
+	// ensure predictable behavior of RandomEarlyDrop.
+	const testSeed = int64(12345)
+	localRng := rand.New(rand.NewSource(testSeed))
+
+	redPred := RandomEarlyDrop[int](
+		minThresh, maxThresh, WithRandSource(localRng.Float64),
+	)
+
+	// Next, we'll define a custom predicate: drop items with value 42.
+	customValuePredicate := func(queueLen int, item int) bool {
+		return item == 42
+	}
+
+	// We'll also make a composed predicate: drop if RED says so OR if item
+	// is 42.
+	composedPredicate := func(queueLen int, item int) bool {
+		isRedDrop := redPred(queueLen, item)
+		isCustomDrop := customValuePredicate(queueLen, item)
+		return isRedDrop || isCustomDrop
+	}
+
+	q := NewBackpressureQueue[int](capacity, composedPredicate)
+
+	// Scenario 1: Enqueue item 42 when queue length is between min/max
+	// thresholds. As we're below the max threshold, we shouldn't drop
+	// anything.
+	for i := 0; i < minThresh; i++ {
+		// All items aren't 42, so they shouldn't be dropped.
+		res := q.Enqueue(context.Background(), i)
+		require.True(
+			t, res.IsOk(), "enqueue S1 setup item %d Ok: %v", i,
+			res.Err(),
+		)
+
+		droppedVal, _ := res.Unpack()
+		require.False(
+			t, droppedVal, "enqueue S1 setup item %d (qLen "+
+				"before: %d) should not be dropped. Predicate "+
+				"was redPred(%d,%d) || customPred(%d,%d)", i,
+			len(q.ch)-1, len(q.ch)-1, i, len(q.ch)-1, i,
+		)
+	}
+
+	currentLen := len(q.ch)
+	require.Equal(t, minThresh, currentLen, "queue length after S1 setup")
+
+	// Enqueue item 42. customValuePredicate is true, so composedPredicate
+	// is true. Item 42 should be dropped regardless of what redPred
+	// decides.
+	res := q.Enqueue(context.Background(), 42)
+	require.True(
+		t, res.IsOk(), "Enqueue of 42 should not error: %v", res.Err(),
+	)
+	require.True(t,
+		res.UnwrapOrFail(t), "Item 42 should have been dropped by "+
```

Comment:
nit. I was confused, because this can never fail, since you checked `res.IsOk()` above. But we are checking the success case of the result, which should be `true`. Maybe consider `res.UnwrapOr(false)`?